### PR TITLE
Feature/create categories collection -- Issue #24

### PR DESCRIPTION
**Pull request by vhazna**
Title: Create Collection Categories

Description:
> We need to:
> Create collection Categories
> So:
> We will use the collection

Ticket link: [ticket link](https://github.com/Sf-62-NodeJS/clothes-market-server/issues/24)

### CODE REVIEW

**Comment by vhazna** [comment]
https://user-images.githubusercontent.com/110221236/188320009-b96ef656-20df-42b9-b603-b537fc314285.mp4

Categories endpoints in Swagger

**Comment by vhazna** [comment]
https://user-images.githubusercontent.com/110221236/188406518-4c2f4b9f-49d2-4d7f-a4ef-25ceadd7ae9d.mp4

Categories create and update -- availability check demo.

File: src/middlewares/validators/updateCategoryPayload.js [hunk 1]
```diff
@@ -0,0 +1,16 @@
+const Joi = require('joi');
+
+const updateCategoryPayloadValidator = (req, res, next) => {
+  const schema = Joi.object({
+    name: Joi.string().min(2).max(30).required()
```
Comment (VasylLymych) [inline src/middlewares/validators/updateCategoryPayload.js]: It there any difference between create and update category payload?

File: src/services/categories.js [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+      return res.boom.badRequest(`Category ${req.body.name} already exists.`);
+    }
+
+    const category = await new Categories(req.body).save();
+
+    return res.json(category);
+  }
+
+  async updateCategory (req, res) {
+    const categoryExists = await Categories.findOne({
```
Comment (VasylLymych) [inline src/services/categories.js]: Code duplication... Create some mutual private method

File: src/services/categories.js [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+      name: req.body.name
+    }).exec();
+
+    if (categoryExists !== null) {
+      return res.boom.badRequest(`Category ${req.body.name} already exists.`);
+    }
+
+    const category = await new Categories(req.body).save();
+
+    return res.json(category);
```
Comment (VasylLymych) [inline src/services/categories.js]: Return true or false. Let's follow our project pattern: create, update, delete methods must return true/false

File: src/middlewares/validators/updateCategoryPayload.js (hunk 1 above)
Comment (vhazna) [reply src/middlewares/validators/updateCategoryPayload.js]: Fixed.

File: src/services/categories.js (hunk 2 above)
Comment (vhazna) [reply src/services/categories.js]: Fixed.

File: src/services/categories.js (hunk 3 above)
Comment (vhazna) [reply src/services/categories.js]: Fixed.

File: src/controllers/categories.js [hunk 4]
```diff
@@ -0,0 +1,22 @@
+const { CategoriesService } = require('../services');
+const categoriesService = new CategoriesService();
```
Comment (VasylLymych) [inline src/controllers/categories.js]: Make this as private field of CategoriesController class (#categoriesService = new CategoriesService())

File: src/middlewares/validators/nameCategoryPayload.js [hunk 5]
```diff
@@ -0,0 +1,16 @@
+const Joi = require('joi');
+
+const nameCategoryPayloadValidator = (req, res, next) => {
```
Comment (VasylLymych) [inline src/middlewares/validators/nameCategoryPayload.js]: What a strange naming? Name it just "categoryPayloadValidator". What if there are more fields, like in user?) Will you name it nameSurnameAge...UserValidator?

File: src/services/categories.js [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+
+    const category = await Categories.findByIdAndUpdate(req.params.id, {
+      name: req.body.name
+    }).exec();
+
+    return category ? res.json(true) : res.boom.notFound();
+  }
+
+  async getCategories (req, res) {
+    const { _id, name } = (typeof req.query !== 'undefined' && req.query) || {};
```
Comment (VasylLymych) [inline src/services/categories.js]: https://github.com/Sf-62-NodeJS/clothes-market-server/pull/44/files#r963763380

File: src/services/categories.js [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+    if (_id != null) query._id = _id;
+    if (name != null) {
+      query.name = { $regex: req.query.name, $options: 'i' };
+    }
+
+    const count = Categories.find(query).count();
+    const category = await Categories.find(query).exec();
+
+    return {
+      total_number: count,
```
Comment (VasylLymych) [inline src/services/categories.js]: {total_size, list}

File: src/middlewares/validators/nameCategoryPayload.js (hunk 5 above)
Comment (vhazna) [reply src/middlewares/validators/nameCategoryPayload.js]: Haha I agree with you, Vasyl. Fixed.

File: src/services/categories.js (hunk 7 above)
Comment (vhazna) [reply src/services/categories.js]: Fixed.

File: src/services/categories.js (hunk 6 above)
Comment (vhazna) [reply src/services/categories.js]: Fixed.

File: src/controllers/categories.js (hunk 4 above)
Comment (vhazna) [reply src/controllers/categories.js]: Applied.

File: src/services/categories.js [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+
+    return category ? res.json(true) : res.boom.notFound();
+  }
+
+  async getCategories (req, res) {
+    const { _id, name } = req.query;
+    const query = {};
+
+    if (_id != null) query._id = _id;
+    if (name != null) {
```
Comment (VasylLymych) [inline src/services/categories.js]: You can just write if(_id)...
if(name)...
No necessity to write if(_id != null)

**Review by VasylLymych — APPROVED** [review]

**Pull request merged**